### PR TITLE
added progressbar and tests

### LINKED ISSUE #85: Add a Progressbar
It would greatly improve the user experience if a progress bar was present during simulation. This could be achieved by using a jax io-callback on a progress bar object.

### PR DESCRIPTION
**Pull request by renaissancenerd**
resolves #85 

---
## Summary by cubic
Adds a JAX-safe progress bar to FDTD runs, enabled by default for reversible, checkpointed, and custom forward modes. Implements #85 with device-gated host updates and no JAX overhead when disabled.

- **New Features**
  - Added `SimulationProgressBar` and helpers (`_auto_update_interval`, `_make_pbar`, `_wrap_body_with_progress`).
  - Device-side gating via `jax.lax.cond` + `jax.experimental.io_callback(ordered=True)`; updates only at intervals; no overhead when the bar is absent.
  - `show_progress` (default True) in `reversible_fdtd`, `checkpointed_fdtd`, `custom_fdtd_forward`; per-mode labels; context-managed bar forces 100% on close.
  - Partial runs use `step_offset` for relative progress; bar skipped when `n_steps=0` or when `start_time`/`end_time` are traced `jax.Array`s (avoid concretization).
  - Probes for `tqdm`; if missing, warn and disable. Tests cover callback order, interval gating, relative progress, traced bounds, and unchanged outputs.
  - Docs: added `notebooks/2D-Simulation.ipynb` to the Advanced guide.

- **Migration**
  - Backwards compatible. `tqdm` is optional; progress shows by default when installed. Disable with `show_progress=False`.

<sup>Written for commit e435171dafab4dc5374ac90e1d3f2c144efd77e4. Summary will update on new commits.</sup>

### CODE REVIEW

File: src/fdtdx/fdtd/fdtd.py [hunk 1]
```diff
@@ -402,6 +551,7 @@ def custom_fdtd_forward(
     record_detectors: bool,
     start_time: int | jax.Array,
     end_time: int | jax.Array,
+    progress_bar: "SimulationProgressBar | None" = None,
```
Comment (ymahlau) [inline src/fdtdx/fdtd/fdtd.py]: please dont use quoted arguments, but rather import `from future import __annotations__` at the top of the file

File: src/fdtdx/fdtd/fdtd.py [hunk 2]
```diff
@@ -13,11 +13,138 @@
...
+        )
+        return self
+
+    def __exit__(self, *_):
+        if self._bar is not None:
+            self._bar.close()
+            self._bar = None
+
+    # Public helpers
+    def reset(self):
```
Comment (ymahlau) [inline src/fdtdx/fdtd/fdtd.py]: Some of these parts can also be deleted when progressbar is only used within the function

File: tests/fdtd/test_fdtd.py [hunk 3]
```diff
@@ -154,3 +156,318 @@ def test_custom_fdtd_forward_same_start_end(dummy_arrays, dummy_objects, dummy_c
...
+
+# Integration tests: progress bar wired into each simulation function
+
+
+class TestProgressBarIntegration:
+    """Verify that passing a SimulationProgressBar to the simulation functions
+    does not break their outputs, and test callback wiring independently of
+    JAX's lax.while_loop effect execution.
+
+    Note: jax.experimental.io_callback effects fired inside jax.lax.while_loop
```
Comment (ymahlau) [inline tests/fdtd/test_fdtd.py]: Is this really true? do you have a reference for that?

**Comment by renaissancenerd** [comment]
> You can add a boolean flag as argument to the functions that optionally disables the progress bar, since updating the progress bar will make computation a little slower.
> 
> On that note, can you make a comparison between the simulation speed with and without progress bar and share it here as a comment please?

── Progress bar overhead benchmark ──────────────────────────────────
  grid: 8³,  steps: 0,  repeats: 3 (excl. warm-up)
  scenario                          best (s)    overhead
  -------------------------------------------------------
  no bar (baseline)                              0.1617s
  auto interval                                      0.1583s  (-2.1 %)
  update_interval=1 ;                            0.1686s  (+4.3 %)
  update_interval=10 ;                          0.1870s  (+15.7 %)
  update_interval=50 ;                          0.1708s  (+5.7 %)
─────────────────────────────────────────────────────────────────────

**Comment by renaissancenerd** [comment]
@l-berg pls also review this PR, as this has been open for a few weeks.
